### PR TITLE
Call base class __init__ in subclasses

### LINKED ISSUE #1207: Initializing base classes
As someone who is coming from C++ I would like to see us calling `super().__init__()` every time we define an initializer in a derived class to ensure proper initialization of the parent's variables. See https://github.com/trezor/trezor-firmware/pull/1180 for a concrete example where this would make sense in my opinion. To me it seems counterintuitive not to initialize the parent by default, but maybe the pythonic way of thinking about this is different? Let's discuss IRL.

### PR DESCRIPTION
**Pull request by mmilata**
Fixes #1207. I tried to copy the pylint's "super-init-not-called (W0231)" approach - child class must call the ancestor's `__init__` provided it has one. Number of allocations across device tests increases by about 0.5 %.

* The changes affect the generated protobuf classes, not sure if there can be impact on performance other than the small amount of allocations.
* Children of `Hash143` do a bit of unnecessary work now, perhaps it would be better to refactor `Hash143.__init__` or add an abstract base class ...

### CODE REVIEW

File: core/src/trezor/ui/__init__.py [hunk 1]
```diff
@@ -306,6 +310,9 @@ class Layout(Component):
     BACKLIGHT_LEVEL = style.BACKLIGHT_NORMAL
     RENDER_SLEEP = loop.sleep(_RENDER_DELAY_MS)  # type: loop.Syscall
 
+    def __init__(self) -> None:
+        super().__init__()
+
```
Comment (andrewkozlik) [inline core/src/trezor/ui/__init__.py:315]: Isn't this implicit?

Comment (mmilata) [reply core/src/trezor/ui/__init__.py:315]: IIUC it is, will remove

**Comment by andrewkozlik** [comment]
It would be nice if we could be running pylint automatically with "super-init-not-called" as part of `make style`, but I guess that's a different issue https://github.com/trezor/trezor-firmware/issues/132.

> The changes affect the generated protobuf classes, not sure if there can be impact on performance other than the small amount of allocations.

Does `MessageType.__init__()` ever get called anywhere with a non-empty set of arguments? If not, then it would make more sense to remove the `__init__` method from `MessageType` rather than add `super().__init__()` to all the derived classes.

> Children of `Hash143` do a bit of unnecessary work now, perhaps it would be better to refactor `Hash143.__init__` or add an abstract base class ...

I agree about adding an abstract base class (or maybe a protocol class?).

**Comment by mmilata** [comment]
> Does `MessageType.__init__()` ever get called anywhere with a non-empty set of arguments? If not, then it would make more sense to remove the `__init__` method from `MessageType` rather than add `super().__init__()` to all the derived classes.

It doesn't look like it's actually used, I ran tests on modified version which throws exception in such case and the tests pass. The [coverage report](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/867414822/artifacts/core/htmlcov/src_protobuf_py.html#n155) is puzzling though.

> I agree about adding an abstract base class (or maybe a protocol class?).

Converted this to protocol in 9045037d7799955d16be54a6ced4c0a04fcb0c5b. Unfortunately I'm OOP-challenged and haven't figured out how to share the code for `add_input` and `add_output` in a way that's not ugly. PTAL

File: core/src/trezor/ui/__init__.py (hunk 1 above)
Comment (mmilata) [reply core/src/trezor/ui/__init__.py]: Fixed in 63e2372c1831791ada10e52f362b6e8f764bc3fa.

**Review by andrewkozlik — APPROVED** [review]
> Converted this to protocol in [9045037](https://github.com/trezor/trezor-firmware/commit/9045037d7799955d16be54a6ced4c0a04fcb0c5b). Unfortunately I'm OOP-challenged and haven't figured out how to share the code for `add_input` and `add_output` in a way that's not ugly. PTAL

I am OK with the current state. If you want to avoid the code duplication, then I think the cleanest way to do it would be to derive `Zip243Hash` from `BitcoinHash143` and create the writers using three factory methods like `self.h_prevouts = self.create_hash_prevouts()` etc. or have one factory method which returns all three like `self.h_prevouts, self.h_sequence, self.h_outputs = self.create_hash_writers()` . Nevertheless I think the code duplication is not that severe and future versions of ZIP243 may diverge further from BIP143, which might break this model anyway.

One thing I would do though is rename `BitcoinHash143` -> `Bip143Hash`.

As for the classes derived from `ui.Component`, I have only looked at those conceptually and I am happy. Someone else might want to have a closer look.

**Review by matejcik — APPROVED** [review]
lgtm

**Pull request merged**